geode-core/src/main/java/org/apache/geode/internal/cache/execute/FunctionContextImpl.java

```diff
@@ -44,19 +45,22 @@
   private ResultSender resultSender = null;
 
   private final boolean isPossDup;
+  private DistributedSystem distributedSystem;
```

Comment:
I don't think we need to store the DS, as the cache already contains the DS

Reply:
Hi @kohlmu-pivotal Thanks for review. There are scenario where in cache is going to be null and that is the reason we need to pass DS explicitly. Consider a test case OnGroupsFunctionExecutionDUnitTest#testBasicP2PFunctionNoCache here function is called on the member and cache is not created.

Reply:
Hi there @davinash. Looking at these tests, they don't seem to be correct. This is because in https://github.com/apache/geode/blob/a237203a3944b1a4daf5162928ca849ab930779c/geode-core/src/test/java/org/apache/geode/internal/cache/execute/OnGroupsFunctionExecutionDUnitTest.java#L119-L146 we create a DistributedSystem using a deprected feature, https://github.com/apache/geode/blob/a237203a3944b1a4daf5162928ca849ab930779c/geode-core/src/test/java/org/apache/geode/test/dunit/internal/JUnit4DistributedTestCase.java#L205 which is to be removed in favour for https://github.com/apache/geode/blob/215177f23119e801e676b8310015841127aa15ef/geode-core/src/main/java/org/apache/geode/cache/CacheFactory.java#L205.
In reality, one would never create a DistributedSystem without a Cache... In addition, you would never directly create a DistributedSystem, you would rather have the CacheFactory deal with all of this.
I'm in favour of removing the offending tests, rather than forcing the FunctionContext to adhere to something that we know will never be the use case.
In addition, I'm looking at moving the FunctionService to be accessed from the cache in a non-static manner.